src/Datadog.Trace/ScopeManager.cs

```diff
@@ -0,0 +1,143 @@
+                            if (activeScope != null)
+                            {
+                                return activeScope;
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            Log.ErrorException($"Active scope access failed: {_prioritizedScopeAccess[i]?.GetType().FullName}.", ex);
+                            continue;
+                        }
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        public Scope Activate(Span span, bool finishOnClose)
+        {
+            var newParent = Active;
+            var scope = new Scope(newParent, span, this, finishOnClose);
+            var scopeOpenedArgs = new SpanEventArgs(span);
+
+            SpanOpened?.Invoke(this, scopeOpenedArgs);
+
+            SetActiveScope(scope);
+
+            if (newParent != null)
+            {
+                SpanDeactivated?.Invoke(this, new SpanEventArgs(newParent.Span));
+            }
+
+            SpanActivated?.Invoke(this, scopeOpenedArgs);
+
+            return scope;
+        }
+
+        public void Close(Scope scope)
+        {
+            var current = Active;
+            var isRootSpan = scope.Parent == null;
+
+            if (current == null || current != scope)
+            {
+                // This is not the current scope for this context, bail out
+                return;
+            }
+
+            // if the scope that was just closed was the active scope,
+            // set its parent as the new active scope
+            SetActiveScope(current.Parent);
+            SpanDeactivated?.Invoke(this, new SpanEventArgs(current.Span));
+
+            if (!isRootSpan)
+            {
+                SpanActivated?.Invoke(this, new SpanEventArgs(current.Parent.Span));
+            }
+
+            SpanClosed?.Invoke(this, new SpanEventArgs(scope.Span));
+
+            if (isRootSpan)
+            {
+                TraceEnded?.Invoke(this, new SpanEventArgs(scope.Span));
+            }
+        }
+
+        public void RegisterScopeAccess(IActiveScopeAccess scopeAccess)
+        {
+            lock (_scopeAccessLock)
+            {
+                if (_prioritizedScopeAccess.Any(i => i.GetType() == scopeAccess.GetType()))
+                {
+                    // We don't want multiple instances as each type manages scope in an ambient fashion.
+                    return;
+                }
+
+                _prioritizedScopeAccess.Add(scopeAccess);
+
+                // Slower add for faster read
+                _prioritizedScopeAccess = _prioritizedScopeAccess.OrderByDescending(i => i.Priority).ToList();
+            }
+        }
+
+        private void SetActiveScope(Scope scope)
+        {
+            lock (_scopeAccessLock)
+            {
+                for (int i = 0; i < _prioritizedScopeAccess.Count; i++)
```

Comment:
Nit: use `foreach`

Reply:
Was a performance choice, maybe not worthwhile?